Allow local databases to be used for kraken2, centrifuge, and busco

gregorysprenger opened the pull request:
This PR is a follow up to issue [#498](https://github.com/nf-core/mag/issues/498). 

This adds the ability to check if a database is compressed or decompressed. Updated BUSCO to only have to specify `--busco_db` for databases that are both compressed or decompressed. It determines if an input is a lineage dataset by looking for `odb10` in the input filename (ie. `bacteria_odb10`). Since I dropped `--busco_reference`, I decided to rename `--save_busco_reference` to `--save_busco_db`. 

<!--
# nf-core/mag pull request

Many thanks for contributing to nf-core/mag!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/mag/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/mag/tree/master/.github/CONTRIBUTING.md)- [ ] If necessary, also make a PR on the nf-core/mag _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

gregorysprenger commented:
> @gregorysprenger I have a little bit of spare time so just had a look: I wonder if the error is coming from the fact that `busco_db` is coming in as a value channel rather than just a path/file object itself? So `isDirectory()` doesn't work on channel?

Yeah, I got it figured out now! Once I finish running the test profiles locally (and they pass), I'll get everything pushed over.

prototaxites commented on lib/WorkflowMag.groovy:
```diff
@@ -102,22 +102,13 @@ class WorkflowMag {
-            }
-            if (params.busco_download_path) {
-                Nextflow.error('Both --skip_binqc and --busco_download_path are specified! Invalid combination, please specify either --skip_binqc or --binqc_tool \'busco\' with --busco_download_path.')
+            if (params.busco_db) {
```
Outwith the scope of this PR for sure, but I think cases like this where you skip a step but specify a database should probably just print a warning rather than quitting. Makes debugging a little easier if you just need to quickly turn something off.

prototaxites commented on nextflow_schema.json:
```diff
@@ -757,23 +757,18 @@
                     "description": "Run BUSCO with automated lineage selection, but ignoring eukaryotes (saves runtime)."
                 },
-                "save_busco_reference": {
+                "save_busco_db": {
```
I'm wondering if this should keep its name (or if it should be renamed to "save_busco_references"? As far as I see it, the use case for this parameter is to save the lineages files downloaded when using online auto lineage detection. On their own, these don't actually comprise a busco database as there are other index files you need to download in order to pass a directory to `busco_db`.

prototaxites commented:
Couple of small comments, but otherwise LGTM I think 👍

jfy133 commented on modules/local/kraken2_db_preparation.nf:
```diff
@@ -14,10 +14,21 @@ process KRAKEN2_DB_PREPARATION {
-    tar -xf "${db}" -C db_tmp
-    mkdir database
-    mv `find db_tmp/ -name "*.k2d"` database/
+    if [[ -d ${db} ]]; then
```
Theoretically we can do the checking if it's a directory with Nextflow's `.isDirectory()` function as well as checkign the name... but this can updated in the future (preferably as making symlinks of symlinks can be dodgy in some computing env contexts)

jfy133 commented on nextflow_schema.json:
```diff
@@ -481,12 +481,12 @@
                     "type": "string",
                     "description": "Database for taxonomic binning with centrifuge.",
-                    "help_text": "E.g. ftp://ftp.ccb.jhu.edu/pub/infphilo/centrifuge/data/p_compressed+h+v.tar.gz."
+                    "help_text": "Local directory containing `*.cf` files or path to download compressed tar archive. E.g. ftp://ftp.ccb.jhu.edu/pub/infphilo/centrifuge/data/p_compressed+h+v.tar.gz."
```
```suggestion
                    "help_text": "Local directory containing `*.cf` files, or a URL or local path to a downloaded compressed tar archive of a Centrifuge database. E.g. ftp://ftp.ccb.jhu.edu/pub/infphilo/centrifuge/data/p_compressed+h+v.tar.gz."
```

jfy133 commented on nextflow_schema.json:
```diff
@@ -481,12 +481,12 @@
                     "type": "string",
                     "description": "Database for taxonomic binning with kraken2.",
-                    "help_text": "The database file must be a compressed tar archive that contains at least the three files `hash.k2d`, `opts.k2d` and `taxo.k2d`. E.g. ftp://ftp.ccb.jhu.edu/pub/data/kraken2_dbs/minikraken_8GB_202003.tgz."
+                    "help_text": "Local directory or compressed tar archive that contains at least the three files `hash.k2d`, `opts.k2d` and `taxo.k2d`. E.g. ftp://ftp.ccb.jhu.edu/pub/data/kraken2_dbs/minikraken_8GB_202003.tgz."
```
```suggestion
                    "help_text": "Path to a local directory, archive file, or a URL to compressed tar archive that contains at least the three files `hash.k2d`, `opts.k2d` and `taxo.k2d`. E.g. ftp://ftp.ccb.jhu.edu/pub/data/kraken2_dbs/minikraken_8GB_202003.tgz."
```

jfy133 commented on nextflow_schema.json:
```diff
@@ -757,23 +757,18 @@
-                    "type": "string",
-                    "description": "Path to local folder containing already downloaded and unpacked lineage datasets.",
-                    "help_text": "If provided, BUSCO analysis will be run in offline mode. Data can be downloaded from https://busco-data.ezlab.org/v5/data/ (files still need to be unpacked manually). Run in combination with automated lineage selection."
+                    "description": "Download path for BUSCO lineage dataset or path to local directory containing already downloaded and unpacked lineage datasets.",
```
```suggestion
                    "description": "Download URL for BUSCO lineage dataset, or path to a tar.gz archive, or local directory containing already downloaded and unpacked lineage datasets.",
```

jfy133 commented on subworkflows/local/busco_qc.nf:
```diff
@@ -9,23 +9,57 @@ include { BUSCO_SUMMARY                   } from '../../modules/local/busco_summ
+                                .map{
+                                    db ->
+                                        def meta = [:]
+                                        meta['id'] = db.toString().split('/').last()
```
I think ti would be safer to use one of the file attribute functions: https://www.nextflow.io/docs/latest/script.html#check-file-attributes

jfy133 commented on subworkflows/local/gtdbtk.nf:
```diff
@@ -24,7 +24,7 @@ workflow GTDBTK {
                         def contamination = -1
                         def missing, duplicated
-                        if (params.busco_reference) {
+                        if (params.busco_db.toString().contains('odb10')) {
```
See above about e.g. `getBaseName` file attribue functions

jfy133 reviewed (commented):
Needs a changelog update! Also given we are changing a parameter it needs tobe loudly flagged (I think thsi will require a 2.5 infact...))

I'm starting my test runs now, hope to report back soon!

jfy133 commented:
Manual Testing results:

## Busco

gregorysprenger commented on modules/local/busco.nf:
```diff
@@ -25,17 +24,16 @@ process BUSCO {
 
-    def p = "--auto-lineage"
-    if (params.busco_reference){
+    def p = params.busco_auto_lineage_prok ? "--auto-lineage-prok" : "--auto-lineage"
```
The `p` variable and `auto-lineage` parameter are both already in the busco module. I just shortened it to a one liner. Also, if I recall correctly, passing in an empty database to busco with `--auto-lineage`, it will automatically download lineages.

I'm not super familiar with the `ext.args` but looking into it, it does make sense. I'm open to suggestions on handling the lineage datasets (full local db vs lineage dataset).

jfy133 replied:
Ah ok... if it was already in there just leave it as it is!

More info on ext.args: https://nf-co.re/docs/contributing/modules#what-are-the-ext-propertieskeys

Although to be honest we should rip out all of these local modules and replace with 'offiical' nf-core ones...

gregorysprenger commented:
@jfy133 Thanks for the suggestions and testing edge cases I overlooked! I'll try to get all of these fixed over the weekend.

jfy133 commented:
I've still got 2 to do, I hope by tomorrow!

jfy133 commented:
OK, it's just the BUSCO uncompressed tar/dir input which is problmatic, and after that we are good to go :+1:

jfy133 commented on CHANGELOG.md:
```diff
@@ -16,6 +16,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - [#481](https://github.com/nf-core/mag/pull/481) - Adds MetaEuk for annotation of eukaryotic MAGs, and MMSeqs2 to enable downloading databases for MetaEuk (by @prototaxites)
 - [#437](https://github.com/nf-core/mag/pull/429) - `--gtdb_db` also now supports directory input of an pre-uncompressed GTDB archive directory (reported by @alneberg, fix by @jfy133)
 - [#494](https://github.com/nf-core/mag/pull/494) - Adds support for saving the BAM files from Bowtie2 mapping of input reads back to assembly (fix by @jfy133)
+- [#504](https://github.com/nf-core/mag/pull/504) - `--busco_db`, `--kraken2_db`, and `--centrifuge_db` now support direcotry input of a pre-uncompressed database archive directory (by @gregorysprenger).
```
```suggestion
- [#504](https://github.com/nf-core/mag/pull/504) - `--busco_db`, `--kraken2_db`, and `--centrifuge_db` now support directory input of a pre-uncompressed database archive directory (by @gregorysprenger).
```

gregorysprenger replied:
Jeez, you'd think I'd stop making the same spelling mistake...

jfy133 commented:
Thanks @gregorysprenger ! I will test again at the weekend :+1:

gregorysprenger commented on modules/local/busco_db_preparation.nf:
```diff
@@ -10,9 +10,8 @@ process BUSCO_DB_PREPARATION {
-    path "buscodb/*"    , emit: db
-    path database       , emit: database
-    path "versions.yml" , emit: versions
+    tuple val("${database.toString().replace(".tar.gz", "")}"), path("buscodb/*"), emit: db
```
Thoughts on doing either:
`tuple val( "${database.getSimpleName()}")...`
 or 
`tuple val( "buscodb")..` ?

jfy133 replied:
The former I guess?

gregorysprenger replied:
Done

gregorysprenger commented:
> Thanks @gregorysprenger ! I will test again at the weekend 👍

Give me a couple more hours and I'll get the centrifuge and kraken database parsing into groovy to avoid the possible symlink issues. :)

gregorysprenger commented:
Alright I think I'm done this time lol. @jfy133 I ran all of your manual tests and they appeared to work fine, but double checking is always best!

Quick updates:
- used groovy to handle if centrifuge/kraken is a compressed file or directory
- updated centrifuge `db_name` parsing. when using centrifuge with the `-x` param, the db_name needs to equal the filename of *.cf files with *.1.cf removed
- changed most of the toString() functions to a nextflow filename attribute function (i e. getBaseName() )

[ its weird that PythonBlack failed O.o ]

prototaxites commented on workflows/mag.nf:
```diff
@@ -463,20 +454,63 @@ workflow MAG {
     */
-    CENTRIFUGE_DB_PREPARATION ( ch_centrifuge_db_file )
+    if ( !ch_centrifuge_db_file.isEmpty() ) {
+        if ( ch_centrifuge_db_file.extension == "gz" || ch_centrifuge_db_file.extension == "tgz" ) {
```
if(ch_centrifuge_db_file.extension in [“gz”, “tgz”]{ ? A bit less verbose

gregorysprenger replied:
Done

jfy133 reviewed (approved):
LGTM, could always bit of refinement here and there but this works :+1: 

One last thing: please pull in latest changes in `dev`, and move your CHANGELOG entry accordingly, then we can merge!

jfy133 commented:
And then I'll od the first patch release :) Thanks @gregorysprenger !

gregorysprenger commented:
Done! Thanks @prototaxites and @jfy133 for reviewing!

jfy133 commented:
Tweaked further the versions to make it clearer the change, I think I will make this `2.5` as this is a larger change.

once tests pass, I will merge and make the release PR :D

jfy133 commented:
@nf-core-bot fix linting
